src/Xamarin.Android.Build.Tasks/Utilities/NdkTools/NdkTools.cs

```diff
@@ -0,0 +1,391 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Runtime.InteropServices;
+
+using Microsoft.Android.Build.Tasks;
+using Microsoft.Build.Utilities;
+using Xamarin.Android.Tools;
+
+namespace Xamarin.Android.Tasks
+{
+	public abstract class NdkTools
+	{
+		// Target triples used in various places (tools prefix, include directory naming etc)
+		static readonly Dictionary<AndroidTargetArch, string> archTriples = new Dictionary<AndroidTargetArch, string> {
+			{ AndroidTargetArch.Arm, "arm-linux-androideabi" },
+			{ AndroidTargetArch.Arm64, "aarch64-linux-android" },
+			{ AndroidTargetArch.X86, "i686-linux-android" },
+			{ AndroidTargetArch.X86_64, "x86_64-linux-android" },
+		};
+
+		// Architecture names as used by the `platforms/*` directories pre NDK r22
+		static readonly Dictionary<AndroidTargetArch, string> archPlatforms = new Dictionary<AndroidTargetArch, string> {
+			{ AndroidTargetArch.Arm, "arm" },
+			{ AndroidTargetArch.Arm64, "arm64" },
+			{ AndroidTargetArch.X86, "x86" },
+			{ AndroidTargetArch.X86_64, "x86_64" },
+		};
+
+		protected Dictionary<NdkToolKind, string> NdkToolNames = new Dictionary<NdkToolKind, string> {
+			{ NdkToolKind.Assembler, "as" },
+			{ NdkToolKind.Linker, "ld" },
+			{ NdkToolKind.Strip, "strip" },
+		};
+
+		public NdkVersion Version { get; }
+		public string NdkRootDirectory { get; }
+		public bool UsesClang { get; protected set; }
+		public bool NoBinutils { get; protected set; }
+
+		// We can't use MonoAndroidHelper.AndroidSdk.AndroidNdkHostPlatform here since it's
+		// not initialized while running tests and attempts to use cause a NREX
+		// We could call `MonoAndroidHelper.RefreshSupportedVersions` but since this NdkTools
+		// instance already knows the location of NDK, it would be just a waste of time.
+		protected string HostPlatform => GetNdkHostPlatform ();
+		protected bool IsWindows => OS.IsWindows;
+
+		protected TaskLoggingHelper? Log { get; }
+
+		protected NdkTools (string androidNdkPath, NdkVersion version, TaskLoggingHelper? log = null)
+		{
+			if (String.IsNullOrEmpty (androidNdkPath)) {
+				throw new ArgumentException ("must be a non-empty string", nameof (androidNdkPath));
+			}
+
+			NdkRootDirectory = androidNdkPath;
+			Version = version;
+		}
+
+		public static NdkTools? Create (string androidNdkPath, TaskLoggingHelper? log = null)
+		{
+			if (String.IsNullOrEmpty (androidNdkPath)) {
+				log?.LogCodedError ("XA5104", Properties.Resources.XA5104);
+				return null;
+			}
+
+			if (!Directory.Exists (androidNdkPath)) {
+				log?.LogCodedError ("XA5104", Properties.Resources.XA5104);
+				return null;
+			}
+
+			NdkVersion? version = ReadVersion (androidNdkPath, log);
+			if (version == null) {
+				return null;
```

Comment:
*May* want to add a `log?.LogDebugMessage("Could not read version information")` or something, if/when this starts returning null…

Reply:
This is kinda/sorta done, actually, as `ReadVersion()` contains some logging.

The "problem"/"concern" is one of "scope" & expectations: since I saw `LogCodedError()` previously in this method, I *expected* a similar log here.  Perhaps my expectation is wrong.

Additionally, `ReadVersion()` has a code path which could throw an exception, which would result in XA5104 *not* being raised, which could be differently underirable.

Reply:
No more exceptions thrown there in the latest update.

Reply:
TBH, I'd much prefer to keep throwing exceptions (also from `.Create`) to have a sorta-kinda RAII-like behavior, and some code somewhere else capture the exceptions and turn them into XA5104, but I don't think there's any central location where that could be done

Reply:
The problem is that the exceptions *are* caught from a central location:

https://github.com/xamarin/xamarin-android-tools/blob/623332da80e1ff2e1a567258c8df8a2a87558aba/src/Microsoft.Android.Build.BaseTasks/AndroidTask.cs#L16-L21

The problem is that they're not transformed into any "controllable" error code; they just go to `Log.LogUnhandledException()`, which makes for a "less than stellar" error experience.

I'm fine with exceptions, as-such.  But I'd *also* like `Log.LogCodedError()`s before the exception, when sane/possible.

Reply:
@grendello  [wrote](https://discord.com/channels/732297728826277939/732297837953679412/864974432242368523):

> I tried this
> exceptions **and*8 a logcodederror
> breaks the tests

*Weird*. :-/